<a href="https://colab.research.google.com/github/younzk/rookie-projects/blob/main/4.8%20Pivot%20Table%20%26%20Group%20Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#4.8 Pivot Table & Group Analysis

##Pivot Table
- Pivot Table이란 데이터 열 중에서 두 개의 열을 각각 행, 인덱스, 열 인덱스로 사용하여 데이터를 조회하여 펼쳐놓은 것
-df1.pivot("행 인덱스로 사용할 열 이름", "열 인덱스로 사용할 열 이름", "데이터로 사용할 열 이름")

- 판다스는 지정된 두 열을 각각 행 인덱스와 열 인덱스로 바꾸고 행 인덱스의 라벨 값이 첫번째 키의 값과 같고 두번째 키의 값과 같은 데이터를 찾아서 해당 칸에 넣는다. 만약 데이터가 존재하지 않으면 NaN값을 넣는다.

In [ ]:
import pandas as pd
import numpy as np
data = {
    "도시": ["서울", "서울", "서울", "부산", "부산", "부산", "인천", "인천"],
    "연도": ["2015", "2010", "2005", "2015", "2010", "2005", "2015", "2010"],
    "인구": [9904312, 9631482, 9762546, 3448737, 3393191, 3512547, 2890451, 263203],
    "지역": ["수도권", "수도권", "수도권", "경상권", "경상권", "경상권", "수도권", "수도권"]
}
columns = ["도시", "연도", "인구", "지역"]
df1 = pd.DataFrame(data, columns=columns)
df1

도시    연도       인구   지역
0  서울  2015  9904312  수도권
1  서울  2010  9631482  수도권
2  서울  2005  9762546  수도권
3  부산  2015  3448737  경상권
4  부산  2010  3393191  경상권
5  부산  2005  3512547  경상권
6  인천  2015  2890451  수도권
7  인천  2010   263203  수도권

In [ ]:
df1.pivot("도시", "연도", "인구") #행인덱스 : 도시, 열인덱스 : 연도, 데이터 : 인구

연도       2005       2010       2015
도시                                 
부산  3512547.0  3393191.0  3448737.0
서울  9762546.0  9631482.0  9904312.0
인천        NaN   263203.0  2890451.0

피봇테이블은 set_index와 unstack 명령을 사용해서 만들 수도 있다.
- set_index : 해당 데이터열을 인덱스로 설정
- unstack : 행 인덱스 -> 열 인덱스

In [ ]:
df1.set_index(["도시", "연도"])[["인구"]]

인구
도시 연도           
서울 2015  9904312
   2010  9631482
   2005  9762546
부산 2015  3448737
   2010  3393191
   2005  3512547
인천 2015  2890451
   2010   263203

In [ ]:
df1.set_index(["도시", "연도"])[["인구"]].unstack()

인구                      
연도       2005       2010       2015
도시                                 
부산  3512547.0  3393191.0  3448737.0
서울  9762546.0  9631482.0  9904312.0
인천        NaN   263203.0  2890451.0

In [ ]:
df1.set_index(["도시", "연도"])[["인구"]].unstack().unstack()

    연도    도시
인구  2005  부산    3512547.0
          서울    9762546.0
          인천          NaN
    2010  부산    3393191.0
          서울    9631482.0
          인천     263203.0
    2015  부산    3448737.0
          서울    9904312.0
          인천    2890451.0
dtype: float64

In [ ]:
df1.pivot(["지역", "도시"], "연도", "인구")

연도           2005       2010       2015
지역  도시                                 
경상권 부산  3512547.0  3393191.0  3448737.0
수도권 서울  9762546.0  9631482.0  9904312.0
    인천        NaN   263203.0  2890451.0

행 인덱스와 열 인덱스는 데이터를 찾는 key역할을 하므로 키 값으로 데이터가 하나면 찾아져야 한다.
행 인덱스와 열 인덱스 조건을 만족하는 데이터가 2개 이상인 경우에는 에러가 발생한다.
예를 들어 위에서 "지역", "연도"를 키로 하면 "수도권", "2015"에 해당하는 값이 두 개 이상이므로 에러가 발생한다.

In [ ]:
try:
  df1.pivot("지역", "연도", "인구")
except ValueError as e:
  print("ValueError: ", e)

ValueError:  Index contains duplicate entries, cannot reshape


##Grourp Analysis(그룹 분석)
- 키가 지정하는 조건에 맞는 데이터가 하나 이상이라 데이터 그룹을 이루는 경우 사용
- groupby 메서드 사용

##그룹 연산 메서드
- size, count : 그룹 데이터의 종류
- mean, median, min, max : 그룹 데이터의 평균, 중앙값, 최소, 최대
- sum, prod, std, var, quantile : 그룹 데이터의 합계, 곱, 표준편차, 분산, 사분위수
- first, last : 그룹 데이터 중 가장 첫번째 데이터와 가장 나중 데이터
- agg, aggregate : 원하는 그룹 연산이 없는 경우 함수를 만들고 이 함수를 agg에 전달, 여러가지 그룹연산을 동시에 하고 싶은 경우 함수 이름 문자열의 리스트 전달
- describe : 여러 개의 값을 데이터 프레임으로 구하기
- apply : describe처럼 하나의 대표값이 아닌 데이터프레임을 출력하지만 원하는 그룹 연산이 없는 경우 사용
- transform : 그룹 대표값을 만드는게 아니라 그룹별 계산을 통해 데이터 자체를 변형

In [ ]:
np.random.seed(0)
df2 = pd.DataFrame({
    'key1': ['A', 'A', 'B', 'B', 'A'],
    'key2': ['one', 'two', 'one', 'two', 'one'],
    'data1': [1, 2, 3, 4, 5],
    'data2': [10, 20, 30, 40, 50]
})
df2

key1 key2  data1  data2
0    A  one      1     10
1    A  two      2     20
2    B  one      3     30
3    B  two      4     40
4    A  one      5     50

key1의 값에 따른 data1의 평균
1. groupby명령을 사용해 그룹 A와 그룹 B로 구분한 그룹데이터를 만든다.

In [ ]:
groups = df2.groupby(df2.key1).mean()
groups

data1      data2
key1                     
A     2.666667  26.666667
B     3.500000  35.000000

GroupBy 클래스 객체에는 각 그룹의 인덱스를 저장한 groups 속성이 있다.

A그룹과 B그룹의 합계를 위해 sum이라는 그룹 연산을 한다.

In [ ]:
groups.sum()

data1     6.166667
data2    61.666667
dtype: float64

GroupBy 클래스가 필요 없다면 groupby메서드와 그룹연산 메서드를 연속 호출한다.

In [ ]:
df2.data1.groupby(df2.key1).sum() #data1에 대해서 key1을 기준으로 그룹 -> sum 연산

key1
A    8
B    7
Name: data1, dtype: int64

In [ ]:
df2.groupby(df2.key1)["data1"].sum() #df2전체를 key1기준으로 그룹 -> 그중 data1 만 -> sum 연산 ('GroupBy) 클래스 객체에서 data1만 선택하여 분석

key1
A    8
B    7
Name: data1, dtype: int64

In [ ]:
df2.groupby(df2.key1).sum()["data1"] #전체 데이터를 분석 후 data1만 선택

key1
A    8
B    7
Name: data1, dtype: int64

복합 키(key1, key2) 값에 따른 data1의 합계 구하기
- 키가 복수인 경우 리스트 사용

In [ ]:
df2.data1.groupby([df2.key1, df2.key2]).sum()

key1  key2
A     one     6
      two     2
B     one     3
      two     4
Name: data1, dtype: int64

unstack을 사용해서 pivot table로 만들 수도 있다.

In [ ]:
df2.data1.groupby([df2["key1"], df2["key2"]]).sum().unstack("key2") #key2(행인덱스)를 열 인덱스로 바꿔준다

key2  one  two
key1          
A       6    2
B       3    4

In [ ]:
df1["인구"].groupby([df1["지역"], df1["연도"]]).sum().unstack("연도")

연도      2005     2010      2015
지역                             
경상권  3512547  3393191   3448737
수도권  9762546  9894685  12794763

150송이의 붓꽃(iris)에 대해 붓꽃 종(species)별로 꽃잎 길이(sepal_length), 꽃잎 폭(sepal_width)을 측정한 데이터 불러오기

In [ ]:
import seaborn as sns
iris = sns.load_dataset("iris")

In [ ]:
iris.head(10)

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
6           4.6          3.4           1.4          0.3  setosa
7           5.0          3.4           1.5          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           1.5          0.1  setosa

각 붓꽃 종별로 가장 큰 값과 가장 작은 값을 비율 구하기
- 메서드가 없으므로 직접 만든 후 agg 메서드를 적용한다.

In [ ]:
def peak_to_peak_ratio(x):
  return x.max() / x.min()

iris.groupby(iris.species).agg(peak_to_peak_ratio) #species로 그룹한 데이터에서 가장 큰 값(max)와 가장 작은 값(min)의 비율

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa          1.348837     1.913043      1.900000     6.000000
versicolor      1.428571     1.700000      1.700000     1.800000
virginica       1.612245     1.727273      1.533333     1.785714

describe 메서드를 사용하면 다양한 기술 통계(descriptive statistics)값을 한 번에 구한다.

In [ ]:
iris.groupby(iris.species).describe().T 

species                setosa  versicolor  virginica
sepal_length count  50.000000   50.000000  50.000000
             mean    5.006000    5.936000   6.588000
             std     0.352490    0.516171   0.635880
             min     4.300000    4.900000   4.900000
             25%     4.800000    5.600000   6.225000
             50%     5.000000    5.900000   6.500000
             75%     5.200000    6.300000   6.900000
             max     5.800000    7.000000   7.900000
sepal_width  count  50.000000   50.000000  50.000000
             mean    3.428000    2.770000   2.974000
             std     0.379064    0.313798   0.322497
             min     2.300000    2.000000   2.200000
             25%     3.200000    2.525000   2.800000
             50%     3.400000    2.800000   3.000000
             75%     3.675000    3.000000   3.175000
             max     4.400000    3.400000   3.800000
petal_length count  50.000000   50.000000  50.000000
             mean    1.462000    4.260000   5.552000
             std     0.173664    0.469911   0.551895
             min     1.000000    3.000000   4.500000
             25%     1.400000    4.000000   5.100000
             50%     1.500000    4.350000   5.550000
             75%     1.575000    4.600000   5.875000
             max     1.900000    5.100000   6.900000
petal_width  count  50.000000   50.000000  50.000000
             mean    0.246000    1.326000   2.026000
             std     0.105386    0.197753   0.274650
             min     0.100000    1.000000   1.400000
             25%     0.200000    1.200000   1.800000
             50%     0.200000    1.300000   2.000000
             75%     0.300000    1.500000   2.300000
             max     0.600000    1.800000   2.500000

apply 메서드를 사용하면 describe 메서드처럼 하나의 그룹에 대해 하나의 대표값(스칼라 값)을 구하는 게 아니라 데이터프레임을 만들 수 있다.  
 예를 들어 다음처럼 각 붓꽃 종별로 가장 꽃잎 길이(petal length)가 큰 3개의 데이터를 뽑아낼 수도 있다.

In [ ]:
def top3_petal_length(df):
    return df.sort_values(by="petal_length", ascending=False)[:3] #sort_values() : 데이터값 기준으로 정렬,  ascending = False :큰 수에서 작은 수로 반대 정렬, by = 기준이 되는 열

k = iris.groupby(iris.species).apply(top3_petal_length)
k

sepal_length  sepal_width  petal_length  petal_width  \
species                                                                
setosa     24            4.8          3.4           1.9          0.2   
           44            5.1          3.8           1.9          0.4   
           23            5.1          3.3           1.7          0.5   
versicolor 83            6.0          2.7           5.1          1.6   
           77            6.7          3.0           5.0          1.7   
           72            6.3          2.5           4.9          1.5   
virginica  118           7.7          2.6           6.9          2.3   
           117           7.7          3.8           6.7          2.2   
           122           7.7          2.8           6.7          2.0   

                   species  
species                     
setosa     24       setosa  
           44       setosa  
           23       setosa  
versicolor 83   versicolor  
           77   versicolor  
           72   versicolor  
virginica  118   virginica  
           117   virginica  
           122   virginica

In [ ]:
k["sepal_length"]["versicolor"]

83    6.0
77    6.7
72    6.3
Name: sepal_length, dtype: float64

transform 메서드는 그룹별 대표값을 만드는 것이 아니라 그룹별 계산을 통해 데이터프레임 자체를 변화시킨다.   
따라서 만들어진 데이터프레임의 크기는 원래 데이터프레임과 같다.   
예를 들어 다음처럼 각 붓꽃 꽃잎길이가 해당 종 내에서 대/중/소 어느 것에 해당되는지에 대한 데이터프레임을 만들 수도 있다.

In [ ]:
def q3cut(s):
    return pd.qcut(s, 3, labels=["소", "중", "대"]).astype(str) #s에 들어오는 값을 3가지 카테고리로 나누기 , astype으로 전체 자료형을 바꾼다.


iris["petal_length_class"] = iris.groupby(iris.species).petal_length.transform(q3cut)
iris[["petal_length", "petal_length_class"]].tail(10)

petal_length petal_length_class
140           5.6                  중
141           5.1                  소
142           5.1                  소
143           5.9                  대
144           5.7                  중
145           5.2                  소
146           5.0                  소
147           5.2                  소
148           5.4                  중
149           5.1                  소

##Practice Test 4. 7. 2
붓꽃(iris) 데이터에서 붓꽃 종(species)별로 꽃잎길이(sepal_length), 꽃잎폭(sepal_width) 등의 평균을 구하라. 만약 붓꽃 종(species)이 표시되지 않았을 때 이 수치들을 이용하여 붓꽃 종을 찾아낼 수 있을지 생각하라.

In [ ]:
iris.head(10)

sepal_length  sepal_width  petal_length  petal_width species  \
0           5.1          3.5           1.4          0.2  setosa   
1           4.9          3.0           1.4          0.2  setosa   
2           4.7          3.2           1.3          0.2  setosa   
3           4.6          3.1           1.5          0.2  setosa   
4           5.0          3.6           1.4          0.2  setosa   
5           5.4          3.9           1.7          0.4  setosa   
6           4.6          3.4           1.4          0.3  setosa   
7           5.0          3.4           1.5          0.2  setosa   
8           4.4          2.9           1.4          0.2  setosa   
9           4.9          3.1           1.5          0.1  setosa   

  petal_length_class  
0                  소  
1                  소  
2                  소  
3                  중  
4                  소  
5                  대  
6                  소  
7                  중  
8                  소  
9                  중

In [ ]:
iris.groupby(iris.species).mean()

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa             5.006        3.428         1.462        0.246
versicolor         5.936        2.770         4.260        1.326
virginica          6.588        2.974         5.552        2.026

##pivot_table
- pivot명령과 groupby 명령의 중간 성격을 가짐
- groupby처럼 그룹 분석을 하지만 최종적으로 pivot처럼 피봇테이블을 만든다.
- groupby 명령의 결과에 unstack을 자동 적용하여 2차원 형태로 변형한다.

* pivot_table(data, values=None, index=None, columns=None, aggfunc='mean', fill_value=None, margins=False, margins_name='All')

 - data: 분석할 데이터프레임 (메서드일 때는 필요하지 않음)

 - values: 분석할 데이터프레임에서 분석할 열

 - index: 행 인덱스로 들어갈 키 열 또는 키 열의 리스트

 - columns: 열 인덱스로 들어갈 키 열 또는 키 열의 리스트

 - aggfunc: 분석 메서드

 - fill_value: NaN 대체 값

 - margins: 모든 데이터를 분석한 결과를 오른쪽과 아래에 붙일지 여부

 - margins_name: 마진 열(행)의 이름

In [ ]:
df1.pivot_table("인구", "도시", "연도")

연도       2005       2010       2015
도시                                 
부산  3512547.0  3393191.0  3448737.0
서울  9762546.0  9631482.0  9904312.0
인천        NaN   263203.0  2890451.0

margins = True 인수 : aggfunc으로 주어진 분석방법을 해당 열의 모든 데이터, 해당 행의 모든 데이터, 그리고 전체 데이터에 대해 적용한 결과를 보여준다. aggfunc이 주어지지 않았으면 평균을 계산한다.

아래 결과에서 가장 오른쪽 합계 열의 첫번째 값 3583992은 모든 부산 인구 데이터의 평균, 두번째 값 9749333은 모든 서울 인구 데이터의 평균이다. 가장 아래의 합계 행의 첫번째 값은 2005년 데이터의 평균값, 두번째 값은 2010년 데이터의 평균값이다. 가장 오른쪽 아래의 값 7283197는 전체 데이터의 평균값이다. 다음 계산을 통해 이를 확인할 수 있다.

In [ ]:
df1.pivot_table("인구", "도시", "연도", margins=True, margins_name="합계") #aggfunc이 없으므로 평균

연도       2005       2010       2015            합계
도시                                               
부산  3512547.0  3393191.0  3448737.0  3.451492e+06
서울  9762546.0  9631482.0  9904312.0  9.766113e+06
인천        NaN   263203.0  2890451.0  1.576827e+06
합계  6637546.5  4429292.0  5414500.0  5.350809e+06

In [ ]:
df1["인구"].mean()

5350808.625

In [ ]:
df1.pivot_table("인구", index = ["연도", "도시"]) #인덱스에 리스트를 넣으면 다중 인덱스 테이블

인구
연도   도시         
2005 부산  3512547
     서울  9762546
2010 부산  3393191
     서울  9631482
     인천   263203
2015 부산  3448737
     서울  9904312
     인천  2890451

In [ ]:
tips = sns.load_dataset("tips")
tips.tail()

total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [ ]:
tips['tip_pct'] = tips['tip'] / tips['total_bill'] #tip_pct : 식사대금 대비 팁의 비율
tips.tail()


total_bill   tip     sex smoker   day    time  size   tip_pct
239       29.03  5.92    Male     No   Sat  Dinner     3  0.203927
240       27.18  2.00  Female    Yes   Sat  Dinner     2  0.073584
241       22.67  2.00    Male    Yes   Sat  Dinner     2  0.088222
242       17.82  1.75    Male     No   Sat  Dinner     2  0.098204
243       18.78  3.00  Female     No  Thur  Dinner     2  0.159744

In [ ]:
tips.describe() #각 열 데이터의 분포 알아보기

total_bill         tip        size     tip_pct
count  244.000000  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672    0.160803
std      8.902412    1.383638    0.951100    0.061072
min      3.070000    1.000000    1.000000    0.035638
25%     13.347500    2.000000    2.000000    0.129127
50%     17.795000    2.900000    2.000000    0.154770
75%     24.127500    3.562500    3.000000    0.191475
max     50.810000   10.000000    6.000000    0.710345

In [ ]:
tips.groupby("sex").count() #성별로 나누어 데이터 갯수 세어보기

total_bill  tip  smoker  day  time  size  tip_pct
sex                                                      
Male           157  157     157  157   157   157      157
Female          87   87      87   87    87    87       87

In [ ]:
tips.groupby("sex").size() #NaN데이터가 없는 경우 더 간단히 표시

sex
Male      157
Female     87
dtype: int64

In [ ]:
tips.groupby(["sex","smoker"]).count()

total_bill  tip  day  time  size  tip_pct
sex    smoker                                           
Male   Yes             60   60   60    60    60       60
       No              97   97   97    97    97       97
Female Yes             33   33   33    33    33       33
       No              54   54   54    54    54       54

In [ ]:
tips.groupby(["smoker", "sex"]).count()

total_bill  tip  day  time  size  tip_pct
smoker sex                                              
Yes    Male            60   60   60    60    60       60
       Female          33   33   33    33    33       33
No     Male            97   97   97    97    97       97
       Female          54   54   54    54    54       54

In [ ]:
tips.pivot_table("tip_pct", "sex", "smoker", aggfunc = "count", margins = True) #피봇테이블 형태로

smoker  Yes   No  All
sex                  
Male     60   97  157
Female   33   54   87
All      93  151  244

In [ ]:
#성별과 흡연 여부에 따른 평균 팁 비율
tips.pivot_table("tip_pct", ["smoker", "sex"]) #aggfunc 이 없으므로 평균 계산

tip_pct
smoker sex             
Yes    Male    0.152771
       Female  0.182150
No     Male    0.160669
       Female  0.156921

여성 흡연자의 팁 비율이 0.18215로 가장 높고, 남성 흡연자의 팁 비율이 0.152771로 가장 낮다. 하지만 이 데이터는 평균을 제외한 분산(variance)등의 다른 통계값이 없으므로 describe 명령으로 여러 통계값을 한 번에 알아본다.

In [ ]:
tips.groupby("sex")[["tip_pct"]].describe()

tip_pct                                                              \
         count      mean       std       min       25%       50%       75%   
sex                                                                          
Male     157.0  0.157651  0.064778  0.035638  0.121389  0.153492  0.186240   
Female    87.0  0.166491  0.053632  0.056433  0.140416  0.155581  0.194266   

                  
             max  
sex               
Male    0.710345  
Female  0.416667

In [ ]:
tips.groupby(["sex", "smoker"])[["tip_pct"]].describe()

tip_pct                                                    \
                count      mean       std       min       25%       50%   
sex    smoker                                                             
Male   Yes       60.0  0.152771  0.090588  0.035638  0.101845  0.141015   
       No        97.0  0.160669  0.041849  0.071804  0.131810  0.157604   
Female Yes       33.0  0.182150  0.071595  0.056433  0.152439  0.173913   
       No        54.0  0.156921  0.036421  0.056797  0.139708  0.149691   

                                   
                    75%       max  
sex    smoker                      
Male   Yes     0.191697  0.710345  
       No      0.186220  0.291990  
Female Yes     0.198216  0.416667  
       No      0.181630  0.252672

##Practice Test 4. 7. 3
팁의 비율이 요일과 점심/저녁 여부, 인원수에 어떤 영향을 받는지 살펴본다.

어떤 요인이 가장 크게 작용하는지 판단할 수 있는 방법이 있는가?

In [ ]:
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct
0       16.99  1.01  Female     No  Sun  Dinner     2  0.059447
1       10.34  1.66    Male     No  Sun  Dinner     3  0.160542
2       21.01  3.50    Male     No  Sun  Dinner     3  0.166587
3       23.68  3.31    Male     No  Sun  Dinner     2  0.139780
4       24.59  3.61  Female     No  Sun  Dinner     4  0.146808

In [ ]:
tips.pivot_table("tip_pct", "time")

tip_pct
time            
Lunch   0.164128
Dinner  0.159518

In [ ]:
tips.pivot_table("tip_pct", "day").sort_values(by = "tip_pct", ascending = False)

tip_pct
day           
Fri   0.169913
Sun   0.166897
Thur  0.161276
Sat   0.153152

In [ ]:
tips.pivot_table("tip_pct", "size").sort_values(by = "tip_pct", ascending = False)

tip_pct
size          
1     0.217292
2     0.165719
6     0.156229
3     0.152157
4     0.145949
5     0.141495

따라서 금요일 점심, 1인 테이블에서 가장 많은 팁을 받을 수 있다.
- 어떤 요인이 크게 작용하는지 판단하는 방법?

In [ ]:
def peak_to_peak(x):
  return x.max() - x.min() #peak_to_peak : 가장 많은 팁과 적은 팁의 차이

tips.groupby(["sex", "smoker"])[["tip"]].agg(peak_to_peak)

tip
sex    smoker      
Male   Yes     9.00
       No      7.75
Female Yes     5.50
       No      4.20

- 여러가지 그룹 연산을 동시에 하고 싶을 때 : 리스트를 사용
- 데이터 열마다 다른 연산을 하고 싶다면 딕셔너리로 넣는다. 
{열 라벨 : 연산 이름 또는 함수}

In [ ]:
#여러가지 그룹 연산을 동시에 하고 싶을 때 : 리스트를 사용
tips.groupby(["sex", "smoker"]).agg(["mean", peak_to_peak])[["total_bill"]]

total_bill             
                    mean peak_to_peak
sex    smoker                        
Male   Yes     22.284500        43.56
       No      19.791237        40.82
Female Yes     17.977879        41.23
       No      18.105185        28.58

In [ ]:
tips.groupby(["sex", "smoker"]).agg(
    {'tip_pct' : 'mean', 'total_bill' : peak_to_peak} #tip_pct열은 평균값, total_bill은 peak_to_peak연산
)

tip_pct  total_bill
sex    smoker                      
Male   Yes     0.152771       43.56
       No      0.160669       40.82
Female Yes     0.182150       41.23
       No      0.156921       28.58

In [ ]:
tips.pivot_table(['tip_pct', 'size'], ['sex', 'day'], 'smoker') #데이터 내용은 tip_pct와 size, 행데이터 인덱스 : sex, day, 열데이터 인덱스 : smoker

size             tip_pct          
smoker            Yes        No       Yes        No
sex    day                                         
Male   Thur  2.300000  2.500000  0.164417  0.165706
       Fri   2.125000  2.000000  0.144730  0.138005
       Sat   2.629630  2.656250  0.139067  0.162132
       Sun   2.600000  2.883721  0.173964  0.158291
Female Thur  2.428571  2.480000  0.163073  0.155971
       Fri   2.000000  2.500000  0.209129  0.165296
       Sat   2.200000  2.307692  0.163817  0.147993
       Sun   2.500000  3.071429  0.237075  0.165710

In [ ]:
tips.pivot_table('size', ['time', 'sex', 'smoker'], 'day', 
                 aggfunc = 'sum', fill_value = 0)

day                   Thur  Fri  Sat  Sun
time   sex    smoker                     
Lunch  Male   Yes       23    5    0    0
              No        50    0    0    0
       Female Yes       17    6    0    0
              No        60    3    0    0
Dinner Male   Yes        0   12   71   39
              No         0    4   85  124
       Female Yes        0    8   33   10
              No         2    2   30   43

##Practice Test 4. 7. 4
- 타이타닉 데이터를 불러온다.

1. qcut 명령으로 세 개의 나이 그룹을 만든다.

2. 성별, 선실, 나이 그룹에 의한 생존율을 데이터프레임으로 계산한다. 행에는 성별 및 나이 그룹에 대한 다중 인덱스를 사용하고 열에는 선실 인덱스를 사용한다. 생존률은 해당 그룹의 생존 인원수를 전체 인원수로 나눈 값이다.

3. 성별 및 선실에 의한 생존율을 피봇 데이터 형태로 만든다.

In [ ]:
titanic = sns.load_dataset("titanic")

In [ ]:
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
titanic["age_group"] = pd.qcut(titanic.age, 3, labels = ['age1', 'age2', 'age3']).astype(str)

In [ ]:
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone age_group  
0    man        True  NaN  Southampton    no  False      age1  
1  woman       False    C    Cherbourg   yes  False      age3  
2  woman       False  NaN  Southampton   yes   True      age2  
3  woman       False    C  Southampton   yes  False      age3  
4    man        True  NaN  Southampton    no   True      age3

In [ ]:
def alive_rate(x):
  return x.sum() / len(titanic)

titanic.groupby(['sex', 'age_group', 'class'])[["survived"]].agg(alive_rate).unstack('class')

survived                    
class                First    Second     Third
sex    age_group                              
female age1       0.023569  0.022447  0.033670
       age2       0.020202  0.033670  0.014590
       age3       0.048260  0.020202  0.004489
       nan        0.010101  0.002245  0.028058
male   age1       0.005612  0.011223  0.019080
       age2       0.012346  0.003367  0.020202
       age3       0.026936  0.002245  0.003367
       nan        0.005612  0.002245  0.010101

In [ ]:
titanic.groupby(['sex', 'class'])[['survived']].agg(alive_rate)

survived
sex    class           
female First   0.102132
       Second  0.078563
       Third   0.080808
male   First   0.050505
       Second  0.019080
       Third   0.052750